### ISSUE TITLE
Corrupted DB.

### ISSUE THREAD

**Issue by fjoesne** [skavanagh/KeyBox#168]
Keybox is running and seems to be fine except that i cannot log into it after a reboot. At first I thought it was a date issue, as this has caused OTP issues before. Tailing keybox.trace.db from the revealed messages such as:

Caused by: org.h2.jdbc.JdbcSQLException: General error: "java.lang.IllegalStateException: Store header is corrupt: nio:/a/keydb/keybox.mv.db [1.4.192/6]" [50000-192]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:345)
        ... 132 more
Caused by: java.lang.IllegalStateException: Store header is corrupt: nio:/a/keydb/keybox.mv.db [1.4.192/6]
        at org.h2.mvstore.DataUtils.newIllegalStateException(DataUtils.java:773)
        at org.h2.mvstore.MVStore.readStoreHeader(MVStore.java:605)
        at org.h2.mvstore.MVStore.<init>(MVStore.java:355)
        ... 127 more

Could this be the cause of failure to log in? seems strange that the keybox starts at all if the DB is faulty.

If a faulty DB-file is the reason, is there a way to check and recover the db? Seems like an awful single point of failure if I have to cut my losses and re-initialize the entire keybox from scratch. I guess It's my fault for not doing a regular backup. There is no evidence for disk-failure, so could the database be corrupted if the computer is shut down (powerloss etc)? .I'm concerned that this may cause problems on one of my more critical environments where i have deployed keybox as well.

**Comment by skavanagh** [comment]
Yes that would definitely cause the failure to login. It's a good idea to always backup the H2 database in jetty/keybox/WEB-INF/classes/keydb and whatever configs you've changed (KeyBoxConfig.properties, log4j.xml, etc..), everything else doesn't matter so much.

This is the first I've heard of this happening.

H2 does have a recovery tool you can try here -
http://www.h2database.com/html/advanced.html#using_recover_tool

Here is the info on how to connect 
https://github.com/skavanagh/KeyBox/blob/master/src/main/java/com/keybox/manage/util/DSPool.java#L70-L72

you'll need that to run the recovery tool

**Comment by skavanagh** [comment]
Also, you may want to try to backup and rename keybox.mv.db to keybox.h2.db in the keydb folder and re-start KeyBox.  It maybe trying to connect to DB as an MV store when it's actually not in that format.

**Comment by fjoesne** [comment]
To time consuming for me to test out the recovery-tool right now (limited java knowledge),  however, I will try it when i get time. I'm just going to re-init keybox on the few servers that were affected.
Is it possible that taking snapshots of the db while the server is running may cause the snapshot to be corrupted; in other words: must the db-file be "closed" to be backed up?

**Comment by skavanagh** [comment]
I've always shut down keybox and copied off the DB.

H2 has utility's for backup too
http://www.h2database.com/html/tutorial.html#upgrade_backup_restore

"Creating a backup by copying the database files while the database is running is not supported, except if the file systems support creating snapshots. With other file systems, it can't be guaranteed that the data is copied in the right order"

What's in the jetty/keybox/WEB-INF/classes/keydb folder ?

B/c I'm not sure your database is corrupted though.  I think your database is in a PageStore format and KeyBox is trying to connect as an MVStore (since the filename is named keybox.mv.db).  You should try and rename to keybox.h2.db then H2 will pick it up as a PageStore database.

**Comment by fjoesne** [comment]
Renaming the db gave me the following: 
` ERROR DBUtils - org.h2.jdbc.JdbcSQLException: Unsupported database file version or invalid file header in file "/a/keydb/keybox.h2.db" [90048-192]
 org.h2.jdbc.JdbcSQLException: Unsupported database file version or invalid file header in file "/a/keydb/keybox.h2.db" [90048-192]
`

**Comment by skavanagh** [comment]
oh.. ok.. so the DB is mounted outside of the docker container? 

Could it be that you are using an older docker image for KeyBox and trying to mount and use a database that was created with a the latest or newer docker image of KeyBox?

I update H2 all the time when I do releases. I'm not sure how an older H2 client would work, with a database created with a newer version of the library.

**Comment by fjoesne** [comment]
The DB is mounted from the docker-host and linked to the correct location inside the container environment. This was not a problem before, so I don't believe docker volumes cause the problem. On the other hand, I should probably consider the way docker stops the container if indeed an improper shutdown can cause wrong state of the DB-file. I will reopen the issue with aditosoftware to explore this possibility in case the issue returns ([https://github.com/aditosoftware/docker-keybox/issues/4]) 

I tested an older version as you suggested , but I can't be certain that its the correct version because I deleted the container-image I used before this happened. This may actually be the cause, I have attempted to recreate the database corruption/"miss-state" but it seems the reason for this remains a mystery :/ Again, I will reopen the issue with aditosoftware in case this happens again.
